flink-contrib/flink-timerserivce-rocksdb/src/main/java/org/apache/flink/contrib/streaming/api/operators/RocksDBInternalTimerService.java

```diff
@@ -0,0 +1,797 @@
+			int startKeyGroup = keyGroupRange.getStartKeyGroup() + (int)(numFatPartitions * (numKeysPerPartition + 1L) + (partitionIndex - numFatPartitions) * numKeysPerPartition);
+			int endKeyGroup = (int)(startKeyGroup + numKeysPerPartition - 1L);
+
+			return (startKeyGroup > endKeyGroup ? null : new KeyGroupRange(startKeyGroup, endKeyGroup));
+		} else {
+			int startKeyGroup = keyGroupRange.getStartKeyGroup() + (int)(partitionIndex * (numKeysPerPartition + 1L));
+			int endKeyGroup = (int)(startKeyGroup + numKeysPerPartition);
+
+			return new KeyGroupRange(startKeyGroup, endKeyGroup);
+		}
+	}
+
+	// ------------------------------------------------------------------------
+	//  Serialization Methods
+	// ------------------------------------------------------------------------
+
+	private byte[] serializeKeyGroup(int keyGroup) {
+		ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+		DataOutputViewStreamWrapper outputView = new DataOutputViewStreamWrapper(outputStream);
+
+		try {
+			IntSerializer.INSTANCE.serialize(keyGroup, outputView);
+		} catch (IOException e) {
+			throw new RuntimeException("Error while deserializing the key group.", e);
+		}
+
+		return outputStream.toByteArray();
+	}
+
+	private byte[] serializeRawTimer(Tuple4<Integer, Long, K, N> rawTimer) {
+		ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+		DataOutputViewStreamWrapper outputView = new DataOutputViewStreamWrapper(outputStream);
+
+		try {
+			IntSerializer.INSTANCE.serialize(rawTimer.f0, outputView);
+			LongSerializer.INSTANCE.serialize(rawTimer.f1, outputView);
+			keySerializer.serialize(rawTimer.f2, outputView);
+			namespaceSerializer.serialize(rawTimer.f3, outputView);
+		} catch (IOException e) {
+			throw new RuntimeException("Error while serializing the raw timer.", e);
+		}
+
+		return outputStream.toByteArray();
+	}
+
+	private Tuple4<Integer, Long, K, N> deserializeRawTimer(byte[] bytes) {
+		ByteArrayInputStream inputStream = new ByteArrayInputStream(bytes);
+		DataInputViewStreamWrapper inputView = new DataInputViewStreamWrapper(inputStream);
+
+		try {
+			int keyGroup = IntSerializer.INSTANCE.deserialize(inputView);
+			long timestamp = LongSerializer.INSTANCE.deserialize(inputView);
+			K key = keySerializer.deserialize(inputView);
+			N namespace = namespaceSerializer.deserialize(inputView);
+
+			return new Tuple4<>(keyGroup, timestamp, key, namespace);
+		} catch (IOException e) {
+			throw new RuntimeException("Error while deserializing the raw timer.", e);
+		}
+	}
+
+	/**
+	 * A timer store backed by RocksDB. 
+	 * 
+	 * The timers are stored in RocksDB in the order of key groups. To allow 
+	 * efficient access, the timers are partitioned and the leader of each 
+	 * partition is stored in an in-memory heap. The top of the heap is 
+	 * exactly the first timer to trigger. The heap is updated whenever the 
+	 * partition's leader is updated.
+	 */
+	private class PersistentTimerHeap {
```

Comment:
I think this class could benefit from a more detailed documentation. For example, about the difference/relationship of key-groups, partitions, fatpartitions.
You could also explain, that this is conceptually like a heap-of-heaps, where the outer heap holds inner heaps, one for each key-group and that the inner heaps hold the timers. Then, the outer heap is in-memory and the inner heaps are in fact completely ordered and held in RocksDB.